### PR TITLE
Jobs shouldn't be reloaded while preparing 

### LINKED ISSUE #7296: reload: preparing jobs may not be flushed pre-reload
When a reload request is received, the scheduler first flushes all preparing jobs through before proceeding to the reload itself.

We have spotted an instance where this did not work, causing preparing jobs to be reloaded, something which Cylc can not handle.

The bug was spotted with 8.6.2. The relevant logic was changed in 8.4.0 which may be a likely cause:

https://github.com/cylc/cylc-flow/commit/68ce6a15e5c868f0f431d65dd78d0d7c3f99a309#diff-aae7aa80aefb65ccc5e18bcb65d8674803cd753e6681c0fe39a697d2f5f1bc14R341

There's no reproduction as yet, though hammering remote submissions & reloads will likely reveal any edge cases fairly easily.

### PR DESCRIPTION
**Pull request by samuel-denton**
Closes #7296
Added a test to reproduce the bug as well as a fix. 
The bug is that the reload flush loop's while condition depends on submit_task_jobs() returning a non-empty list, but tasks in remote-init can hit a continue branch that skips adding them to the return list, causing the loop to exit while tasks are potentially still preparing due to async operations.  

The fix is to also check tasks reported state and waiting_on_job_prep status before exiting the reload flush loop.

The test works by simulating a task taking longer than one reload flush loop to complete remote-init, catching the situation where release_tasks_to_run() returns False (because there were no tasks ready for submission) but a task is still waiting_on_job_prep or in the preparing state. 

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [ ] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Review by oliver-sanders — COMMENTED** [review]
V nice.

File: cylc/flow/commands.py [hunk 1]
```diff
@@ -528,7 +528,10 @@ async def reload_workflow(schd: 'Scheduler', reload_global: bool = False):
 
     # flush out preparing tasks before attempting reload
     schd.reload_pending = 'waiting for pending tasks to submit'
-    while schd.release_tasks_to_run():
+    while schd.release_tasks_to_run() or any(
+        itask.waiting_on_job_prep
+        for itask in schd.pool.get_tasks()
+    ):
```
Comment (oliver-sanders) [inline cylc/flow/commands.py:534]: I think this loop should keep spinning as long as there are any tasks with `waiting_on_job_prep == True` OR tasks in the `preparing` state.

The latter might be already covered by `release_tasks_to_run`, not sure, but should probs check.

Comment (samuel-denton) [reply cylc/flow/commands.py:534]: Good spot, I think that is a gap.  I've updated to:
```py
    while schd.release_tasks_to_run() or any(
        itask.waiting_on_job_prep or itask.state(TASK_STATUS_PREPARING)
        for itask in schd.pool.get_tasks()
    ):
```

File: tests/integration/test_reload.py [hunk 2]
```diff
@@ -372,6 +372,78 @@ async def test_orphan_reload(
         )
 
 
+async def test_reload_waits_for_preparing_tasks_in_remote_init(
```
Comment (oliver-sanders) [inline tests/integration/test_reload.py]: FYI: There is already a test covering the PREPARING state (lines 35-111).

But that test didn't catch the `waiting_on_job_prep` (e.g, stuck in remote-init) scenario.

Would suggest moving this test just under to keep them together.

**Review by oliver-sanders — APPROVED** [review]
LGTM.

Asserted that the new test failed before, but passes after.

File: tests/integration/test_reload.py (hunk 2 above)
Comment (samuel-denton) [reply tests/integration/test_reload.py]: Ah yeah, makes sense to group them.  Have done so.

**Review by MetRonnie — APPROVED** [review]
Looks good.

(N.B. for mistakes like these

Image

we generally prefer it if you squash locally and force-push, rather than leave messy commit history)
